### ISSUE TITLE
Don't add inotify watches on project libraries.

### ISSUE THREAD

**Issue by christianhomberg** [rstudio/rstudio#3267]
One of my favorite RStudio features is sometimes broken.

### System details

    RStudio Edition : Desktop
    RStudio Version : 1.1.442
    OS Version      : Ubuntu 18.04
    R Version       : 3.4.4

### Steps to reproduce the problem
Open a project
### Describe the problem in detail
Sometimes after opening a project the following error appears in the console:

    Warning message:
    File monitoring failed for project at "~/Documents/Projects/mk_waste"
    Error 28 (No space left on device)
    Features disabled: R source file indexing, Diagnostics

The log in `.rstudio-desktop/log/rsession-user.log` is:

    06 Aug 2018 10:52:25 [rsession- user] ERROR system error 28 (No space left on device) [path=/home/ user/.rstudio-desktop/monitored]; OCCURRED AT: rstudio::core::Error rstudio::core::system::file_monitor::{anonymous}::addWatch(const rstudio::core::FileInfo&, const rstudio::core::FilePath&, bool, int, rstudio::core::system::file_monitor::{anonymous}::Watches*) /home/ubuntu/rstudio/src/cpp/core/system/file_monitor/LinuxFileMonitor.cpp:226; LOGGED FROM: void rstudio::session::module_context::{anonymous}::onMonitoringError(const rstudio::core::Error&) /home/ubuntu/rstudio/src/cpp/session/SessionModuleContext.cpp:507
### Describe the behavior you expected
No error to appear when projects are opened and go to file/function to work. At the moment this does work sometimes but I can't tell what the pattern is.

Needless to say that I have plenty of space on my /home drive (> 5GB). Restarting RStudio never works as far as I can tell. Rebooting Ubuntu does.

I'd include a diagnostic report but that would contain file paths which I'd rather not disclose. All files are local.

**Comment by christianhomberg** [comment]
Cheers, I'll keep an eye on this but can see how this works.

I assume we should leave this open because the error prompted in the RStudio console should be different?

**Comment by kevinushey** [comment]
Perhaps -- we're just forwarding on the system error, which (unhelpfully) is the plain old "no space left on device" error you're seeing. I would agree that it would be best if we provided our own more accurate error description.

**Comment by PMassicotte** [comment]
I also have this issue (I have 60%+ disk space):

![image](https://user-images.githubusercontent.com/4519221/56828875-46f9b680-6830-11e9-91df-c1a72ff4672d.png)

![image](https://user-images.githubusercontent.com/4519221/56828948-75779180-6830-11e9-90ac-559abb2fa9c5.png)

**Comment by slhck** [comment]
I'm running into the same issue. Rstudio seems to be the only program affected though – but once the error occurs, the whole system grinds to a halt. The files affected are all related to packrat, by the way.

```
Jul 15 14:16:57 voltaire rsession-werner[31542]: ERROR system error 28 (No space left on device) [path=/home/werner/Documents/Projects/data-analyze/rscript/packrat/src/xml2]; OCCURR
Jul 15 14:16:57 voltaire rsession-werner[31542]: ERROR system error 28 (No space left on device) [path=/home/werner/Documents/Projects/data-analyze/rscript/packrat/src/xopen]; OCCUR
Jul 15 14:16:57 voltaire rsession-werner[31542]: ERROR system error 28 (No space left on device) [path=/home/werner/Documents/Projects/data-analyze/rscript/packrat/src/yaml]; OCCURR
Jul 15 14:16:57 voltaire rsession-werner[31542]: ERROR system error 28 (No space left on device) [path=/home/werner/Documents/Projects/data-analyze/rscript/packrat/src/zeallot]; OCC

```

Could it be that packrat and/or Rstudio are opening such watches and never closing them?

I don't have many open files in my R project, maybe a dozen.

**Comment by mspittler** [comment]
I am actually running into the same error on Linux Mint 19.2 and RStudio 1.2.5019

**Comment by kevinushey** [comment]
Did you read my comment above re: adjusting the number of inotify watches on your system?

**Comment by slhck** [comment]
I had tried this before and it didn't help. It's like increasing RAM if there's a memory leak: it won't help in the long run.

**Comment by mspittler** [comment]
Excatly, sorry, but this did not help me either

**Comment by kevinushey** [comment]
Marking as v1.3 as we should ensure that we don't add inotify watches on project libraries.

**Comment by pgramme** [comment]
First of all, best wishes to all of you!

Got the same error here, with the same path as OP. Please ignore this message if the cause is already clear for you.

I had never got the error before, it began 2 days ago, after an update of R and a crash of RStudio while loading. Now I get the error consistently, even right after reboot when only RStudio is running (+all usual background services) and a small Shiny project is open (120 files in folder and subfolder, no packrat, no project-specific libPath). I can provide you more info and logs if it helps.

System: Ubuntu 18.04, r-base-core 3.6.2-1bionic, RStudio Desktop 1.1.456

**Comment by pgramme** [comment]
**Update**: I just noticed that Spyder (python IDE) also displays the same warning when it tries to do an `inotify_add_watch(...)`. Actually the lack of inotify watches available seems a global bug in Ubuntu, not related to having RStudio running.

Other affected users can test it from a terminal: 
    
    sudo apt install inotify-tools
    # reboot and check whether next line gives an error:
    inotifywatch -t 5 ~/

I did observe an error on my system

**Comment by jmcphers** [comment]
Note: need to ask renv what folder this is. 

    renv:::renv_paths_library

**Comment by t-kalinowski** [comment]
I too am running into this on Ubuntu 18.04 after upgrading to Rstudio 1.2.5033. No packrat or renv usage. Merely adding an empty project specific `.Rprofile` is sufficient to trigger the bug.

**Comment by pgramme** [comment]
RStudio reports the lack of inotify watches, which is a system-wide error. But RStudio may or may not be the culprit. If you run into this error, you should check [this script](https://github.com/fatso83/dotfiles/blob/master/utils/scripts/inotify-consumers) which will give you the number of watches consumed by each process. In my case, it was Kite, the auto-completion daemon for Spyder (python IDE).

**Suggestion:** Error message could be more clear, maybe including the number of watches consumed by current RStudio process.

**Comment by florianm** [comment]
@pgramme thanks for the script! The script showed that in my case sublime took up too many inodes.

```
  INOTIFY                                          
  WATCHER                                                                                               
  COUNT   PID     CMD                                                                                
----------------------------------------                                                                 
   65017   6702  /opt/sublime_text/plugin_host 6687 --auto-shell-env                                   
   65017   6687  /opt/sublime_text/sublime_text                                                        
     139     3309  /usr/lib/gnome-settings-daemon/gsd-xsettings                                          
...
```

Solution:

* Close sublime
* Restart R session

<details>

```
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 19.10

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C               LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8     LC_MONETARY=en_AU.UTF-8   
 [6] LC_MESSAGES=en_AU.UTF-8    LC_PAPER=en_AU.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] usethis_1.5.1.9000

loaded via a namespace (and not attached):
[1] compiler_3.6.1   tools_3.6.1      fs_1.3.1         glue_1.3.1       Rcpp_1.0.3       packrat_0.5.0    rlang_0.4.2.9001

```

</details>

**Comment by nlarusstone** [comment]
I'm still seeing this issue when running `rocker/verse:4.0.3` [Docker Image](https://github.com/rocker-org/rocker-versioned2/blob/master/dockerfiles/Dockerfile_verse_4.0.3). Here's the output of [the counting script](https://github.com/fatso83/dotfiles/blob/master/utils/scripts/inotify-consumers):
```
   INOTIFY
   WATCHER
    COUNT     PID USER     COMMAND
--------------------------------------
    8191   33864 rstudio  /usr/lib/rstudio-server/bin/rsession -u rstudio --session-use-secure-cookies 0 --session-root-path / --session-same-site 0 --launcher-token E81AC932 --r-restore-workspace 2 --r-run-rprofile 2

    8191  WATCHERS TOTAL COUNT
```

I *can* update the number of allowed watchers, but I'm running a number of these docker containers encountering this issue and it seems like this is a bug, so I'd prefer not to have to do a hacky workaround on all of them.

**Comment by kevinushey** [comment]
Can you confirm what files RStudio has actually placed its watches on?

8191 is a very low number of watches so I would encourage you to increase it.

**Comment by nlarusstone** [comment]
`8192` is the default in Linux (I believe), so I would contend that this is a real issue that affects a lot of users.

Can you point me to a snippet that would let me get file names for inotify watches? I was only able to find out how to figure out what process they're associated with, not what path.

**Comment by kevinushey** [comment]
I see this on my Ubuntu 20.04 Docker container:

```
root@0916aca3dd4d:/# cat /proc/sys/fs/inotify/max_user_watches
524288
```

And CentOS 8:

```
[root@3fafa66d3313 /]# cat /proc/sys/fs/inotify/max_user_watches
524288
```

These are pretty vanilla.

That said, you're right that it's hard to ascertain what files are actually marked with inotify watches. The most likely explanation is that there are a large number of files in the open project though, since I think we add a watch for each file in the project? If so, you might be able to confirm by checking the number of watches in a plain / empty R project.

**Comment by AltfunsMA** [comment]
FWIW, I can confirm various VMs I've got running both Ubuntu 18.04 and 20.04 have 8192 `max_user_watches` by default. I'd never come across this until today though; and the culprit in my case is most certainly the fact that I created an Rstudio project at the top level of the attached volume with all my data from a handful of different projects. 

It is also a common issue for VS Code as well, and the [officially proposed solutions](https://code.visualstudio.com/docs/setup/linux#_visual-studio-code-is-unable-to-watch-for-file-changes-in-this-large-workspace-error-enospc) include increasing max_user_watches or excluding folders, which perhaps Rstudio could also provide.
